### PR TITLE
feat(TDP-3419): remove duplicated code in lookup grid services

### PR DESCRIPTION
**Pull request by charlesnguyen**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-3419

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [x] The new code does not introduce new technical issues (sonar / eslint)
- [x] Functional tests have been performed

**Please check the browsers you've tested on**
- [x] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

**(Optional) What is the current behavior?**
(Additional information to the Jira)

**(Optional) What is the new behavior?**
(Additional information to the Jira)

**(Optional) Other information**:

### CODE REVIEW

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/782/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/782/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/789/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/789/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/790/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/790/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/793/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/793/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/796/) [:star:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/796/)

File: dataprep-webapp/src/app/components/datagrid/datagrid-directive.js [hunk 1]
```diff
@@ -45,7 +45,6 @@ export default function Datagrid($timeout, state, DatagridGridService, DatagridC
 		controllerAs: 'datagridCtrl',
 		controller() {
 			this.state = state;
-			this.datagridTooltipService = DatagridTooltipService;
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/datagrid-directive.js]: Did you think about removing the Service injection?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-column-service.js [hunk 2]
```diff
@@ -122,16 +126,16 @@ export default function DatagridColumnService($rootScope, $compile, $log, $trans
...
-	function createColumns(columnsMetadata, preview) {
+	function createColumns(columnsMetadata, preview, colIndexNameTemplate) {
 		// create new SlickGrid columns
 		const colIndexArray = [];
 
 		// Add index column
 		colIndexArray.push({
 			id: COLUMN_INDEX_ID,
-			name: '',
+			name: colIndexNameTemplate ? colIndexNameTemplate : '',
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-column-service.js]: `name: colIndexNameTemplate || '',` looks more readable

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.js [hunk 3]
```diff
@@ -148,27 +144,16 @@ export default function DatagridTooltipService($timeout, state, TextFormatServic
...
      */
-	function init(newGrid) {
-		grid = newGrid;
-		attachTooltipListener();
+	function init(grid, gridState) {
+		// show tooltip on hover
+		grid.onMouseEnter.subscribe((event) => updateTooltip(event, grid, gridState));
+
+		// hide tooltip on leave
+		grid.onMouseLeave.subscribe((event) => hideTooltip(gridState));
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.js]: No eslint error here? `grid.onMouseLeave.subscribe(event => hideTooltip(gridState));`

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.js [hunk 4]
```diff
@@ -148,27 +144,16 @@ export default function DatagridTooltipService($timeout, state, TextFormatServic
...
      * @methodOf data-prep.datagrid.service:DatagridTooltipService
      * @param {object} newGrid The new grid
      * @description Initialize the grid and attach the tooltips listeners
      */
-	function init(newGrid) {
-		grid = newGrid;
-		attachTooltipListener();
+	function init(grid, gridState) {
+		// show tooltip on hover
+		grid.onMouseEnter.subscribe((event) => updateTooltip(event, grid, gridState));
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.js]: No eslint error here? `grid.onMouseEnter.subscribe(event => updateTooltip(event, grid, gridState));`

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 5]
```diff
@@ -47,7 +47,7 @@ describe('Datagrid tooltip service', function () {
     };
 
     //columns metadata
-    var columns = [
+    let columns = [
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 6]
```diff
@@ -134,28 +125,28 @@ describe('Datagrid tooltip service', function () {
...
-            var cell = { row: 1, cell: 0 };                           // contains 'tata'
+            let box = { left: 400, right: 500, top: 10, bottom: 40 }; //width: 100
+            let cell = { row: 1, cell: 0 };                           // contains 'tata'
             gridMock.initCellMock(cell, box);
 
 
             //when
-            var mouseEnterHandler = gridMock.onMouseEnter.subscribe.calls.argsFor(0)[0];
-            var event = { clientX: 500, clientY: 300 };
+            let mouseEnterHandler = gridMock.onMouseEnter.subscribe.calls.argsFor(0)[0];
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 7]
```diff
@@ -134,28 +125,28 @@ describe('Datagrid tooltip service', function () {
...
+            let event = { clientX: 500, clientY: 300 };
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 8]
```diff
@@ -165,108 +156,108 @@ describe('Datagrid tooltip service', function () {
...
 
-        it('should show trailing spaces in the tooltip (column 1)', inject(function ($timeout, DatagridTooltipService) {
+        it('should show trailing spaces in the tooltip (column 1)', inject(($timeout, DatagridTooltipService) => {
             //given
-            DatagridTooltipService.init(gridMock);
+            DatagridTooltipService.init(gridMock, stateMock.playground.grid);
 
-            var box = { left: 400, right: 500, top: 10, bottom: 40 }; //width: 100
-            var cell = { row: 1, cell: 1 }; // contains '  tetetetetetetetetetetetetetetetetetetete '
+            let box = { left: 400, right: 500, top: 10, bottom: 40 }; //width: 100
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 9]
```diff
@@ -165,108 +156,108 @@ describe('Datagrid tooltip service', function () {
...
+            let cell = { row: 1, cell: 1 }; // contains '  tetetetetetetetetetetetetetetetetetetete '
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 10]
```diff
@@ -165,108 +156,108 @@ describe('Datagrid tooltip service', function () {
...
             gridMock.initCellMock(cell, box);
 
             //when
-            var mouseEnterHandler = gridMock.onMouseEnter.subscribe.calls.argsFor(0)[0];
-            var event = { clientX: 500, clientY: 300 };
+            let mouseEnterHandler = gridMock.onMouseEnter.subscribe.calls.argsFor(0)[0];
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js [hunk 11]
```diff
@@ -165,108 +156,108 @@ describe('Datagrid tooltip service', function () {
...
+            let event = { clientX: 500, clientY: 300 };
```
Comment (frassinier) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-tooltip-service.spec.js]: const?

File: dataprep-webapp/src/app/components/datagrid/services/datagrid-column-service.js (hunk 2 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/datagrid/services/datagrid-column-service.js]: done

File: dataprep-webapp/src/app/components/datagrid/datagrid-directive.js (hunk 1 above)
Comment (charlesnguyen) [inline dataprep-webapp/src/app/components/datagrid/datagrid-directive.js]: done

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/807/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/807/)

**Comment by build-talend-tdp** [comment]
Pull request test result: [failure](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/812/) [:x:](https://ci-tdp.datapwn.com/job/tdp_pull-request_master/812/)

**Pull request merged**